doc-level monitor API - findings index and search API (#604)
opensearch-project/documentation-website: _monitoring-plugins/alerting/api.md

# Alerting API

> Use the Alerting API to programmatically create, update and manage monitors and alerts.

Reviewer: ```suggestion
Use the Alerting API to programmatically create, update, and manage monitors and alerts.
```

---

## Create a per query monitor
Introduced 1.0
{: .label .label-purple }

> Per query monitors run the query and check whether the results should trigger any alerts. As such, per query monitors can only trigger one alert at a time. For more information about per query monitors and per bucket monitors, see [Create monitors]({{site.url}}{{site.baseurl}}/monitoring-plugins/alerting/monitors/#create-monitors).

Reviewer: ```suggestion
Per query monitors run the query and check whether the results should trigger an alert. Note that per query monitors can only trigger one alert at a time. For more information about per query monitors and per bucket monitors, see [Create monitors]({{site.url}}{{site.baseurl}}/monitoring-plugins/alerting/monitors/#create-monitors).
```

  }
}
```
## Per document monitors
Introduced 2.0
{: .label .label-purple }

> Per document monitors check whether or not individual documents in an index match trigger conditions to generate an alert notification. When you run a query with a per document monitor, the results are returned for each document that matches the trigger condition. You can create trigger conditions based on query names, query IDs, or tags that combine multiple queries. 

Reviewer: ```suggestion
Per document monitors check whether individual documents in an index match trigger conditions. If so, an alert notification is generated. When you run a query with a per document monitor, the results are returned for each document that matches the trigger condition. You can create trigger conditions based on query names, query IDs, or tags that combine multiple queries. 
```

The following table shows the syntax to use for each trigger option:

Trigger options | Definition | Syntax
:--- | :--- : :---

> Tag | Creates alerts for documents that match a multiple query with this tag applied. If you group multiple queries by a single tag, then you can set it to trigger an alert if the results are returned by this tag name.| `query[tag=<tag-name>]`

Reviewer: This is a little clunky, but not sure how to fix it. : )

Query by name | Creates alerts for documents matched or returned by the named query.  | `query[name=<query-name>]`
Query by ID | Creates alerts for documents that were returned by the identified query. | `query[id=<query-id>]`

For more information about per document monitors, see [Monitor types]({{site.url}}{{site.baseurl}}/monitoring-plugins/alerting/monitors/#monitor-types).

### Search for monitor findings

You can search the findings index `.opensearch-alerting-finding*` for available document findings with a GET request to the findings resource:

```json
GET /_plugins/_alerting/findings/_search?
```

> The response returns all findings available.

Reviewer 2: Can a finding be unavailable? Or does it just return every finding in that system index? If so, I would change this to, "By default, the request returns all findings."

To retrieve metadata for an individual document finding entry, you can search for the finding by its `findingId` as follows:

```json
GET /_plugins/_alerting/findings/_search?findingId=gKQhj8WJit3BxjGfiOXC
```

The response returns the number of individual finding entries in the `total_findings` field.

To get more specific results in a findings search, you can use the path parameters defined in the following table:

Path parameter | Description | Usage
:--- | :--- : :---
`findingId` | The identifier for the finding entry. | The finding ID is returned in the initial query response.

> `sortString` | This field sorts the finding. | The default value is `id`.

Reviewer 2: Instead of "This field sorts the finding," maybe we can say, "This field specifies what string the Alerting plugin uses to sort findings."

`sortOrder` | The order to sort the list of findings, either ascending or descending. | Use `sortOrder=asc` to indicate ascending, or `sortOrder=desc` for descending sort order.

> `size` | The maximum number of results to retrieve. | Can be any integer.

Reviewer 2: Is there a default integer or any minimum/maximum?

`startIndex` | The pagination indicator. | Default is `0`.

> `searchString` | The finding attribute you want returned in the search. | You might want to search findings for a specific index, provide `searchString=indexABC'.

Reviewer 2: "To search in a specific index, specify the index's name in your request's path. For example, ..."

### Create a per document monitor

> You can create a per document monitor by calling the `monitors` resource and providing the monitor details in the body of the request.

Reviewer 2: Is there any wording we can use here instead of "calling `monitors` resource"? Maybe we can say something like, "submit a POST request with the monitor's details in the request's body."

Author: sure

{: .label .label-purple }

Per document monitors check whether individual documents in an index match trigger conditions. If so, the monitor generates an alert notification. When you run a query with a per document monitor, the results are returned for each document that matches the trigger condition. You can create trigger conditions based on query names, query IDs, or tags that combine multiple queries.

The following table shows the syntax to use for each trigger option:

Trigger options | Definition | Syntax
:--- | :--- : :---
Tag | Creates alerts for documents that match a multiple query with this tag applied. If you group multiple queries by a single tag, then you can set it to trigger an alert if the results are returned by this tag name.| `query[tag=<tag-name>]`
Query by name | Creates alerts for documents matched or returned by the named query.  | `query[name=<query-name>]`
Query by ID | Creates alerts for documents that were returned by the identified query. | `query[id=<query-id>]`

For more information about per document monitors, see [Monitor types]({{site.url}}{{site.baseurl}}/monitoring-plugins/alerting/monitors/#monitor-types).

### Search for monitor findings

> You can search the findings index `.opensearch-alerting-finding*` for available document findings with a GET request. By default, a GET request without path parameters returns all available findings.. To learn more about monitor findings, see [Document findings]({{site.url}}{{site.baseurl}}/monitoring-plugins/alerting/monitors/#document-findings).

Reviewer 2: Extra period after "all available findings."
